datafusion/expr/src/expr.rs

```diff
@@ -81,8 +81,10 @@ use std::sync::Arc;
 ///   assert_eq!(binary_expr.op, Operator::Eq);
 /// }
 /// ```
-#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Default)]
 pub enum Expr {
+    #[default]
```

Comment:
Can you explain a bit about the need / usecase for `Expr::Nop`? Could the same be accomplished with  `Option<Expr>`?

Reply:
The reason why I added `Expr::Nop` is to be able to refactor `unalias()` to be self mutating one: https://github.com/apache/arrow-datafusion/pull/7942/files#diff-204cfc4f999c3d12dc065f323cb952fb0ecb33c5570eed8dc1fb52b806e87004R960. I needed a dummy `Expr` for `mem::take()`, but as you showed in https://github.com/apache/arrow-datafusion/pull/7942/files#r1431799873 it doens't need to be self mutating.

But, maybe having a default dummy value of `Expr` is still useful in some cases, like in @sadboy's PR: https://github.com/apache/arrow-datafusion/pull/8591/files#diff-6515fda3c67b9d487ab491fd21c27e32c411a8cbee24725d737290d19c10c199R388-R389, https://github.com/apache/arrow-datafusion/pull/8591/files#diff-6515fda3c67b9d487ab491fd21c27e32c411a8cbee24725d737290d19c10c199R427-R428, `Expr::Wildcard { qualifier: None }` is used for such purposes.

Reply:
Yes, that was my intention. `Noop` sounds good to me.

Reply:
> needed a dummy Expr

If it's just for this purpose, then I think the `Null` literal should serve it well enough:
```
impl Default for Expr {
    fn default() -> Self {
        Expr::Literal(ScalarValue::Null)
    }
}
```

As @alamb mentioned above, there is a (relatively high) cost to introducing new Expr variants, as it increases potential invalid states along every step of the analyzer/optimizer pipeline.

Reply:
Using `Null` literal sounds good to me.

Reply:
Fixed in https://github.com/apache/arrow-datafusion/pull/7942/commits/6cd5d39bfd4eb17adc139103f996fe834b26fbf2.